contrib/rabbitmq/pkg/adapter/adapter.go

```diff
@@ -0,0 +1,223 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rabbitmq
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/cloudevents/sdk-go/pkg/cloudevents"
+	"github.com/cloudevents/sdk-go/pkg/cloudevents/client"
+	"github.com/cloudevents/sdk-go/pkg/cloudevents/types"
+	"github.com/knative/eventing-sources/pkg/kncloudevents"
+	"github.com/knative/pkg/logging"
+	"github.com/streadway/amqp"
+	"go.uber.org/zap"
+    "strings"
+)
+
+const (
+	eventType = "dev.knative.rabbitmq.event"
+)
+
+type ExchangeConfig struct {
+	Name        string
+	TypeOf      string
+	Durable     bool
+	AutoDeleted bool
+	Internal    bool
+	NoWait      bool
+}
+
+type QueueConfig struct {
+	Name             string
+	RoutingKey       string
+	Durable          bool
+	DeleteWhenUnused bool
+	Exclusive        bool
+	NoWait           bool
+}
+
+type Adapter struct {
+	Brokers        string
+	Topic          string
+	ExchangeConfig ExchangeConfig
+	QueueConfig    QueueConfig
+	SinkURI        string
+	client         client.Client
+}
+
+func (a *Adapter) initClient() error {
+	if a.client == nil {
+		var err error
+		if a.client, err = kncloudevents.NewDefaultClient(a.SinkURI); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (a *Adapter) Start(ctx context.Context, stopCh <-chan struct{}) error {
+	logger := logging.FromContext(ctx)
+
+	logger.Info("Starting with config: ", zap.Any("adapter", a))
+
+	conn, err := amqp.Dial(a.Brokers)
+	if err != nil {
+		panic(err)
+	}
+	defer conn.Close()
+
+	ch, err := conn.Channel()
+	if err != nil {
+		panic(err)
+	}
+	defer ch.Close()
+
+	exchangeConfig := fillDefaultValuesForExchangeConfig(&a.ExchangeConfig, a.Topic)
+
+	err = ch.ExchangeDeclare(
+		exchangeConfig.Name,
+		exchangeConfig.TypeOf,
+		exchangeConfig.Durable,
+		exchangeConfig.AutoDeleted,
+		exchangeConfig.Internal,
+		exchangeConfig.NoWait,
+		nil,)
+	if err != nil {
+		panic(err)
+	}
+
+	queue, err := ch.QueueDeclare(
+		a.QueueConfig.Name,
+		a.QueueConfig.Durable,
+		a.QueueConfig.DeleteWhenUnused,
+		a.QueueConfig.Exclusive,
+		a.QueueConfig.NoWait,
+		nil,)
+	if err != nil {
+		panic(err)
+	}
+
+	routingKeys := strings.Split(a.QueueConfig.RoutingKey, ",")
+
+	for _, routingKey := range routingKeys {
+		err = ch.QueueBind(
+			queue.Name,
+			routingKey,
+			exchangeConfig.Name,
+			a.QueueConfig.NoWait,
+			nil,)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	if err := a.initClient(); err != nil {
```

Comment:
If that fail did RabbitMQ register Queue and the Binding? I think that can be good to move this at the top.